ggml/src/ggml-cuda/mmf.cuh

```diff
@@ -57,31 +57,37 @@ static __global__ void mul_mat_f(
     T * tile_xy = (T *) compute_base + threadIdx.y*(tile_A::I * tile_k_padded);
 
     if constexpr (has_ids) {
-        __shared__ int has_any;
-        if (threadIdx.y == 0) {
-            int local_has_any = 0;
-            for (int j = threadIdx.x; j < cols_per_block; j += warp_size) {
-                int slot = -1;
-                for (int k = 0; k < nchannels_dst; ++k) {
-                    const int idv = ids[j*stride_row_id + k*stride_col_id];
-                    if (idv == expert_idx) {
-                        slot = k;
-                        break;
+        int found = 0;
+
+        for (int j = threadIdx.y; j < cols_per_block; j += nwarps) {
```

Comment:
This is a bad pattern because the compiler cannot unroll the loop. You should do it like this instead:

```
#pragma unroll
for (int j0 = 0; j0 < cols_per_block; j0 += nwarps) {
    const int j = j0 + threadIdx.y;
}
```